drivers: iio: ti-dac5571.c

espiriki opened the pull request:
Added the chip_id to the of_device_id array.
Changed the probe function of the driver to use that information (if available)
and set the chip_id accordingly

pinchartl committed:
media: i2c: imx296: Read sensor temperature at probe time

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: IMX296 camera sensor driver

The IMX296LLR is a monochrome 1.60MP CMOS sensor from Sony. The driver
supports cropping and binning (but not both at the same time due to
hardware limitations) and exposure, gain, vertical blanking and test
pattern controls.

Preliminary support is also included for the color IMX296LQR sensor.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

Mani-Sadhasivam committed:
dt-bindings: media: i2c: Add IMX296 CMOS sensor binding

Add YAML devicetree binding for IMX296 CMOS image sensor. Let's also
add MAINTAINERS entry for the binding and driver.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

pinchartl committed:
media: i2c: imx296: Fix power on

Increase the power count early when powering on to ensure controls get
applied.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Fix test pattern configuration

When disabling the test pattern, but IMX296_PGCTRL_CLKEN needs to be
kept, otherwise incorrect behaviour occurs and the output image is
garbage.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Fix test pattern selection

The test pattern selector starts at 0, while the menu control uses value
0 to disable the test pattern generator. Fix it.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Set opaque background in test patterns

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: Add driver for Vision Components MIPI modules

Vision Components manufacture camera modules with different sensors. All
modules have a microcontroller that controls power to the sensor. Add a
minimal driver to support it. The driver should possibly be extended
later to support dynamic identification of the sensor.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Initialize format field and colorspace

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Reorder subdev pad operations

Match the order of operations in the v4l2_subdev_pad_ops structure.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Expose crop bounds and defaults

The crop bounds and defaults are needed by libcamera. Expose them.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Create controls for fwnode properties

Create the V4L2_CID_ORIENTATION and V4L2_CID_ROTATION controls to
expose the corresponding fwnode properties.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Add V4L2_CID_HBLANK control

Hardcode the control value for now, it will be recomputed based on the
format later.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Replace gain control with analogue gain

libcamera expects V4L2_CID_ANALOGUE_GAIN, not V4L2_CID_GAIN.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
dt-bindings: media: i2c: imx296: Add model-specific compatible strings

The IMX296 sensor exists in two different models, a colour variant
(IMX296LQ) and a monochrome variant (IMX296LL). The device exposes the
model through registers, allowing for auto-detection with a common
"sony,imx296" compatible string. However, some camera modules disable
the ability to read the sensor model register, which disables this
feature. In those cases, the exact model needs to be specified as
"sony,imx296ll" or "sony,imx296lq".

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: imx296: Support forcing model through compatible string

The IMX296 sensor exists in two different models, a colour variant
(IMX296LQ) and a monochrome variant (IMX296LL). The device exposes the
model through registers, allowing for auto-detection with a common
"sony,imx296" compatible string. However, some camera modules disable
the ability to read the sensor model register, which makes
auto-detection impossible.

The DT bindings support model-specific "sony,imx296ll" and
"sony,imx296lq" compatible strings. When those are used, skip the
auto-detection.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Sort I2C nodes by address

The usual practice is to sort I2C device nodes by address. Do so. No
functional change included.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add overlay for IMX327 camera module on CSI2

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Drop #address-cells and #size-cells from i2c3

The #address-cells and #size-cells property are already specified for
i2c3 in imx8mp.dtsi. Drop them here.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Drop unneeded labels

Labels are only needed when referenced. Drop unneeded ones.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Drop camera support

The camera module isn't a part of the Maivin board, but a pluggable
component. It should be modelled in DT overlays.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Drop the trigger regulator

The trigger signal to the camera isn't related to a power supply. Drop
the corresponding regulator.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

espiriki committed:
arm64: dts: imx8mp-maivin: Enable the EEPROM on the carrier board

The Maivin carrier board has an EEPROM on bus i2c4 at address 57. It is
already described in the device tree, but isn't functional due to two
issues:

- The imx8mp-verdin.dtsi base device tree already defines a DT node for
  the EEPROM, with status set to disabled. The node that is defined in
  imx8mp-maivin.dts only overrides properties of the base node, without
  overridding the status.

  This behaviour is confusing, and hardware on the carrier board should
  likely not be defined in the base DT, but given that situation, it's
  best to follow how imx8mp-versin.dtsi was meant to be used, and enable
  the eeprom_carrier_board device.

- The i2c4 bus is disabled by default and needs to be enabled.

Fix both issues.

Signed-off-by: Jose Cazarin <joseespiriki@gmail.com>
Reviewed-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Sort top-level nodes alphabetically

The usual practice is to sort top-level nodes alphabetically. Do so. No
functional change included.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Rename regulator-csi to csi-3v3-regulator

This follows the schematics naming, which is the usual practice.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Fix typo

uart1 is used in RS485 mode, not RS245.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Replace "enabled" status with "okay"

The standard status for operational devices is "okay", not "enabled".
Fix it.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add startup delay for csi-3v3 regulator

The csi-3v3 regulator needs time to turn on. The right startup time
needs to be determined, use a value that seems to work for now.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Rename nodes according to device function

The device tree standard recommends naming DT nodes based on the
device's function and not the precise device model. Do so.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: Add device tree source for the Au-Zone Maivin carried board

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add overlay for IMX296 camera module on CSI2

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add overlay for OV5640 camera module on CSI1

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Fix white space

Indent with tabs, fix alignment, add spaces after colon and around equal
sign, adjust comment style.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Adapt to the latest imx7-mipi-csis bindings

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add overlay for IMX296 camera module on CSI1

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Remove unneeded status properties

The status of device tree nodes is "okay" by default when not specified.
Remove unneeded explicit status properties.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Fix the rear connector interface eeprom

The rear connector interface eeprom is connected to bus "I2C_1", which,
confusingly, is i2c4. Move it to the right bus and define it fully.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: vc-mipi: Add clock support

The VC controller provides a clock to the sensor, register a
corresponding clock device.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: vc-mipi: Add structure for driver private data

The driver will need to store private data, add a new vc_mipi_ctrl
structure.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: vc-mipi: Add support for supply regulator

The VC module controller needs a power supply. Acquire a corresponding
regulator. The regulator is kept enabled at all times for now.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: vc-mipi: Retrieve the clock frequency from DT

Not all modules use the same clock frequency. The frequency is
unfortunately not reported by the module ROM, so we have to specific it
in DT.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Rename CSI regulator node

Rename the CSI 3V3 regulator node to start with the standard "regulator"
prefix.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Increase the CSI regulator startup delay

Failures to power up camera modules have been noticed. Increase the
regulator delay. The delay still has to be properly characterized.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Add overlay for IMX327 camera module on CSI1

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Enable M.2 card interface

Enable the PCIe controller, PCIe PHY and the USB_2 controller for the
M.2 interface. The power to the interface is supplied by a
GPIO-controlled regulator, model it as a fixed regulator.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Don't include imx8mp-verdin-dev.dtsi

imx8mp-verdin-dev.dtsi models a development add-on to the Dahlia carrier
board. Maivin isn't based on Dahlia, don't include
imx8mp-verdin-dev.dtsi.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Enable USB recovery interface

Enable the USB_1 controller for the USB recovery interface.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Enable eqos and uart3 devices

The eqos and uart3 are used by Maivin for ethernet and the Linux console
respectively. Enable them.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Enable i2c2 and i2c3 devices

The i2c2 and i2c3 buses have child devices on the Maivin carrier board
itself.  Enable them already to ensure consistent numbering of I2C
buses.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
arm64: dts: imx8mp-maivin: Enable SD card slot

Enable the SD_1 controller for the SD card slot.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
media: i2c: vc-mipi: Make the clock name unique

The common clock framework requires clock names to be unique. Use the
device name to ensure this.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>

pinchartl committed:
Merge branches 'trdx/v5.4.115/sensors', 'trdx/v5.4.115/csis', 'trdx/v5.4.115/isi' and 'trdx/v5.4.115/dt' into trdx/v5.4.115/merge

pinchartl commented on arch/arm64/boot/dts/freescale/imx8mp-maivin.dts:
```diff
@@ -68,11 +68,23 @@
+		compatible = "ti,dac081c081", "ti,dac5571";
+		reg = <0x0d>;
+		vref-supply = <&csi_3v3>;
+	};
```
The DT part looks good, but it should be split to a separate commit. The general rule is that there should be one commit per functional change. Combining the changes to the DAC driver and the Maivin device tree in the same commit would make it impossible to apply one of the changes without the other, or to send the changes to the DAC driver for upstream review for instance.

pinchartl commented on drivers/iio/dac/ti-dac5571.c:
```diff
@@ -387,7 +396,7 @@ static int dac5571_remove(struct i2c_client *i2c)
 #ifdef CONFIG_OF
 static const struct of_device_id dac5571_of_id[] = {
-	{.compatible = "ti,dac5571"},
+	{.compatible = "ti,dac5571", .data = (void *)single_8bit},
```
.data should be set for all entries, otherwise the code in probe() will crash in the OF case for the other devices.

pinchartl commented on drivers/iio/dac/ti-dac5571.c:
```diff
@@ -328,7 +329,15 @@ static int dac5571_probe(struct i2c_client *client,
 	indio_dev->channels = dac5571_channels;
 
-	spec = &dac5571_spec[id->driver_data];
+	if (dev->of_node){
```
The kernel coding style requires a space before {. The } before the else should be on the same line as the else, with a space between the two. But in any case the curly braces should be removed, as the kernel coding style doesn't use them for single-line statements.

I recommend getting familiar with scripts/checkpatch.pl, it's a script that takes a patch file (which you can generate with `git format-patchl`) and checks for lots of common issues (covering both coding style violations and functional issues).

pinchartl reviewed (changes requested):
Your Signed-off-by line is missing. You can add it automatically with the -s option to git commit.

Commit:
arm64: dts: imx8mp-maivin: Add nodes to represent the DAC's on the board

Add DT nods that represents the two DAC's present on the Maivin board

Signed-off-by: Jose Cazarin <joseespiriki@gmail.com>

Commit:
iio: dac: dac5571: Fix chip id detection for OF devices

When matching an OF device, the match mechanism tries all components of
the compatible property. This can result with a device matched with a
compatible string that isn't the first in the compatible list. For
instance, with a compatible property set to

    compatible = "ti,dac081c081", "ti,dac5571";

the driver will match the second compatible string, as the first one
isn't listed in the of_device_id table. The device will howver be named
"dac081c081" by the I2C core.

This causes an issue when identifying the chip. The probe function
receives a i2c_device_id that comes from the module's I2C device ID
table. There is no entry in that table for "dac081c081", which results
in a NULL pointer passed to the probe function.

To fix this, add chip_id information in the data field of the OF device
ID table, and retrieve it with of_device_get_match_data() for OF
devices.

Signed-off-by: Jose Cazarin <joseespiriki@gmail.com>

pinchartl commented on drivers/iio/dac/ti-dac5571.c:
```diff
@@ -328,7 +329,14 @@ static int dac5571_probe(struct i2c_client *client,
-	spec = &dac5571_spec[id->driver_data];
+	if (dev->of_node)
+		chip_id = (uintptr_t)of_device_get_match_data(dev);
+
```
There's an extra blank line here.

pinchartl commented on drivers/iio/dac/ti-dac5571.c:
```diff
@@ -387,15 +395,15 @@ static int dac5571_remove(struct i2c_client *i2c)
-	{.compatible = "ti,dac5573"},
-	{.compatible = "ti,dac6573"},
-	{.compatible = "ti,dac7573"},
+	{.compatible = "ti,dac5571", .data = (void *)single_8bit },
```
Nitpicking, as there's no space after {, there should be none before }. I'd personally add a space in both locations, but when modifying existing drivers, it's best to stick to the existing coding style.

pinchartl reviewed (changes requested):
On the code side; there are just two minor issues in the ti-dac5571.c driver.

The commit message of the patch touching that driver also needs some improvements. First of all, the subject line is currently `drivers: iio: ti-dac5571.c`, which doesn't summarize what the commit does. If you look at the git log for that file, you'll see the following:

```
f46ac009780c iio: dac: Convert powerdown read callbacks to sysfs_emit()
16723c6eaa9e iio:dac:ti-dac5571: Drop of_match_ptr and CONFIG_OF protections
5dfb88af3d52 iio: dac: dac5571: Support powerdown for multi-channel
3593cd53962f Replace HTTP links with HTTPS ones: drivers/iio
8cb631ccbb19 iio: Remove superfluous of_node assignments
d3be83244c7d iio: remove explicit IIO device parent assignment
d2912cb15bdd treewide: Replace GPLv2 boilerplate/reference with SPDX - rule 500
d43102f525d9 iio: dac: ti-dac5571: Update the module author email address
91d05d7639f3 iio: dac: ti-dac5571: provide of_match_table to driver
366995df8b9e iio: dac: ti-dac5571 remove redundant variable 'shift'
df38a4a72a3b iio: dac: add TI DAC5571 family support
```

Let's thus use `iio: dac: dac5571:` as a prefix. If you're wondering why some commits just use `iio:` as a prefix, it's because those commits perform refactoring in the iio subsystem in multiple drivers at a time, and thus have a more generic prefix.

The commit message line could thus be something like `iio: dac: dac5571: Fix chip id detection for OF devices`.

For the body of the commit message, you currently have

```
Adding the chip id to the data field of the of_device_id array.
Changed the probe function to use that information if it is available
```

This describes what the commit does, but not why, which is more important. For simple commits at least, what a commit does is usually easy to figure out by reading the code changes, while the reason behind the change may not be immediately apparent. I'd propose the following commit message.

```
iio: dac: dac5571: Fix chip id detection for OF devices

When matching an OF device, the match mechanism tries all components of 
the compatible property. This can result with a device matched with a
compatible string that isn't the first in the compatible list. For
instance, with a compatible property set to

    compatible = "ti,dac081c081", "ti,dac5571";

the driver will match the second compatible string, as the first one 
isn't listed in the of_device_id table. The device will howver be named
"dac081c081" by the I2C core.

This causes an issue when identifying the chip. The probe function 
receives a i2c_device_id that comes from the module's I2C device ID
table. There is no entry in that table for "dac081c081", which results
in a NULL pointer passed to the probe function.

To fix this, add chip_id information in the data field of the OF device 
ID table, and retrieve it with of_device_get_match_data() for OF
devices.
```

pinchartl reviewed (commented):
s/nods/nodes/ in the commit message. I'll fix this when pushing.

pinchartl commented:
The changes have been merged in the devel branch.
